explorer/interpreter/type_checker.cpp

```diff
@@ -2765,6 +2788,14 @@ auto TypeChecker::TypeCheckStmt(Nonnull<Statement*> s,
       } else {
         CARBON_RETURN_IF_ERROR(TypeCheckPattern(
             &var.pattern(), std::nullopt, var_scope, var.value_category()));
+
+        const auto& value = var.pattern().static_type();
```

Comment:
I'm not clear that there should be any logic change here: @geoffromer might be more familiar with the pattern handling than myself, but I'd expect something like this to be handled somewhere within TypeCheckPattern. Handling it at the statement level doesn't address tuple destructuring cases, e.g.:

```cpp
var (x: auto, y: [i32;]) = (1, (1, 2, 3));
```

Reply:
I see, I'll wait for @geoffromer for more detailed change suggestion.

Reply:
As @jonmeow pointed out elsewhere, a type like `[i32;]` is closely analogous to `auto`, so the logic for inferring the actual size, or raising an error if it can't be inferred, should really be the same as the logic for inferring the actual type of an `auto` binding. That logic starts [here](https://github.com/carbon-language/carbon-lang/blob/45a5828bb43538ee73632aa80645b0c855b695f0/explorer/interpreter/type_checker.cpp#L2743), in `TypeCheckPattern`.

At that point in the code, we're typechecking a binding pattern, and `type` represents the value of the binding pattern's type subpattern. So in a binding pattern like `x: auto`, `type` represents the pattern `auto`. If the binding pattern is in a context where we know the type of the value it's being matched with, that type is stored in `expected`, so this branch basically overwrites `type` with that expected type, after making sure the expected type matches the `type` pattern. Then, after the branch, we verify that `type` is now a _concrete_ type, rather than a pattern that matches types, like `auto`.

So I think there are basically two changes we need to make here:
- We need to change that code to overwrite `type` correctly in the array case. We can't just blindly use `expected` because that's going to be a tuple type.
- We need to update `IsType` to treat a `StaticArrayType` with no size the same way it treats `auto`: as a type but not a concrete type.

Does that make sense?

Reply:
Your explanation is pretty clear, just one more thing to ask, in `IsType`, I should return false if it's concrete type its size is unknown?

Reply:
Yes, I think geoffromer is saying `if (concrete) return type.size().has_value();`

Reply:
@geoffromer It seems that `IsType` was called before type deduction in the beginning of `IsImplicitlyConvertible`, should it be there? The current algorithm is to manually create another `StaticArrayType` instance which its size is infered from `destination`, but `IsType` gets the un-deducted type so it always failed.

Reply:
Implicit conversions can only be defined for concrete types. In other words, considering types A and B, an implicit conversion needs to convert into a specific call to A.ImplicitAs(B), which can only be defined for two concrete types A and B.

In the case of auto, IsType is returning !concrete, so I believe the overall behavior is correct here. This probably instead indicates that there is still space to make the approach you're doing match `auto` more closely, so that IsImplicitlyConvertible is instead called with the deducted type.

I can't comment in more specifics -- I'm about to go away for a couple weeks, but hopefully the above explanation is helpful.